rfcs/0127-issues-warnings.md

```diff
@@ -0,0 +1,283 @@
+  maintainerless = "ignore";
+};
+```
+
+It may also be expanded by values from other configuration options as part of a migration scheme from the other mechanisms.
+
+## Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+### Propagation across transitive dependencies
+
+When a package has a problem that `throw`s, all packages that depend on it will fail to evaluate until that problem is ignored or resolved. Most of the time, this is sufficient.
+
+When the problem requires actions on dependents however, it does not sufficiently inform about all packages that need action. Multiple packages may be annotated with the same problem, in that case it should be given a name and the name should be the same across all instances. Other values like the message or the URL list do not need to be the same and may be adapted sensibly.
+
+For the example of Python 2 deprecation, all problems would have `name = "python2-eol"` and then a user may set `config."*".python2-eol = "ignore";` to ignore them.
+
+### Backwards compatbility, Backporting and stable releases
+
+New problems generally should not be added to stable branches if possible, and also not be backported to them, since it may break evaluation. The same rule applies to other changes to a pacakge's `meta` which may generate a problem and thus lead to evaluation failure too. Scenarios where evaluation failure is a desired goal, for example with unfixable security issues, are obviously exempt from this.
+
+### Removal of packages
+
+The plan is to eventually remove packages with long outstanding problems. The details will be part of future work, but at the very least a package must have a problem whose kind defaults to "throw" for at least one full release cycle (so that stable users have sufficient time to be warned and intervene).
+
+If a package needs to be removed for some other reason, the problem kind `removal` should be used instead:
+
+```nix
+meta.problems = [{
+  kind = "removal";
+  message = "We don't want this in nixpkgs anymore";
+  date = "1970-01-01";
+}];
+```
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+- Too much warnigns may cause Alarm fatigue
+  - One idea I had is to be more verbose on the unstable channels, and then tune down the noise after branch-off.
+  - New lints to packages should be introduced gradually, by making them "silent" by default on start and only going to "warn" after most problems in nixpkgs itself are resolved.
+- People have voiced strong negative opinions about the prospect of removing packages from nixpkgs at all, especially when they still *technically* work.
+  - We do not want to encourage the use of unmaintained software likely to contain security vulnerabilities, and we do not have the bandwidth to maintain packages deprecated by upstream. Nothing is lost though, because we have complete binary cache coverage of old nixpkgs versions, providing a comparatively easy way to pin old very package versions.
```

Comment:
> We do not want to encourage the use of unmaintained software

This "unmaintained" boogeyman is a problem endemic to poorly-scoped software.  Cough cough browsers cough.

`dnscache`, `tinydns`, and `runit` have been "unmaintained" for two decades and work better than most of their competitors.  Is this RFC going to affect their presence in nixpkgs?

Reply:
This is not the kind of "unmaintained" the sentence is referring to. Yes, eventually people may start to enforce `maintainers != []` using the features provided by this RFC if they want to, but this will be a new and separate discussion for sure.

This sentence is about all a) packages that are *effectively* unmaintained, i.e. they require some work to keep going but there is nobody and b) packages that are unmaintained or abandoned *by upstream* with no hope of fixing.

And no, "unmaintained" does not imply it's insecure by any means. Personally, I'd like to get to a point where we treat "end of life" as kind-of insecure. And this is not only about security, but also about workload: I do not want us to patch things on software if it is end of life, instead we should have a graceful way of removing it from `nixpkgs`.